Add connection pooling for HTTP connections

abhin4v opened the pull request:
- connection pool is kept in the ConnectionData data.
- initConnectionData(Secure) take ConnectionPoolConfig as a parameter now.
- withConnection function now ges the connection from the pool instead of creating a new connection each time
- sets AutoClose to `No`
- added mmCloseSession function which closes the connection pool

Fixes issue #43.

abhin4v committed:
Adds connection pooling for HTTP connections.

- connection pool is kept in the ConnectionData data.
- withConnection function now ges the connection from the pool
- added mmCloseSession function which closes the connection pool

jtdaugherty commented on src/Network/Mattermost/Types.hs:
```diff
@@ -49,38 +53,56 @@ maybeFail :: Parser a -> Parser (Maybe a)
+
+maxConnCount :: Int
+maxConnCount = 5
+
```
Can you please make a new record data type containing these settings, such as `data PoolSettings`, provide a `defaultPoolSettings` with these values, and then make that a parameter to `initConnectionData(Insecure)`?

jtdaugherty commented on mattermost-api.cabal:
```diff
@@ -42,6 +42,7 @@ library
                      , aeson >= 1.0.0.0
                      , connection
                      , memory <0.14.3
+                     , resource-pool
```
Can you please add a lower bound for this dependency? (The latest released major version is okay.)

jtdaugherty commented on src/Network/Mattermost/Types/Internal.hs:
```diff
@@ -33,11 +30,12 @@ autoCloseToHeader Yes = [mkHeader HdrConnection "Close"]
-  , cdToken         :: Maybe Token
-  , cdLogger        :: Maybe Logger
-  , cdUseTLS        :: Bool
+  { cdHostname       :: Hostname
```
In the future, please make whitespace-only changes in separate commits.

jtdaugherty commented on src/Network/Mattermost/WebSocket.hs:
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE RecordWildCards #-}
```
Also, bear in mind that we tend not to use this extension much unless the benefit is significant. (Here it is minimal.)

jtdaugherty commented:
Thanks for these changes! I've posted some comments. Also, @aisamanra is working on APIv4 changes to `mattermost-api` and will need to weigh in on whether this can be merged soon or need to wait.

abhin4v committed:
Creates ConnectionPoolConfig to encapsulate connection pool configs.

- defaultConnectionPoolConfig contains the default values for the configs.
- initConnectionData(Secure) take ConnectionPoolConfig as a parameter now.

abhin4v committed:
Fixes examples and tests.

abhin4v commented:
@jtdaugherty I've made the changes you suggested.

jtdaugherty commented:
Thanks!

jtdaugherty committed:
Merge branch 'develop' into feature/connection-pooling
